endpoints-control/src/main/java/com/google/api/control/ControlFilter.java

```diff
@@ -263,30 +268,48 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
 
       // 'Send' a report, end the latency timer to collect correct overhead and backend latencies
       timer.end();
-      ReportRequest reportRequest =
-          createReportRequest(info, checkInfo, appInfo, ConfigFilter.getReportRule(request), timer);
-      if (log.isLoggable(Level.FINEST)) {
-        log.log(Level.FINEST, String.format("sending an error report request %s", reportRequest));
-      }
-      client.report(reportRequest);
-
+      boolean trickle = false;
       if (errorInfo == CheckErrorInfo.API_KEY_NOT_PROVIDED) {
         // a needed API key was not provided
         HttpServletResponse httpResponse = (HttpServletResponse) response;
         httpResponse.sendError(errorInfo.getHttpCode(), errorInfo.getMessage());
       } else if (checkResponse == null) {
         // the check did not complete: 'fail open'
-        chain.doFilter(request, response);
+        trickle = true;
       } else {
         // the checked failed: assume that any error information will be in the first check error
         HttpServletResponse httpResponse = (HttpServletResponse) response;
         httpResponse.sendError(errorInfo.getHttpCode(),
             errorInfo.fullMessage(projectId, checkResponse.getCheckErrors(0).getDetail()));
       }
-      statistics.totalFiltered.incrementAndGet();
-      statistics.totalFilteredTime.addAndGet(overallTimer.elapsed(TimeUnit.MILLISECONDS));
-      logStatistics();
-      return;
+      if (!trickle) {
+        ReportRequest reportRequest =
+            createReportRequest(info, checkInfo, appInfo, ConfigFilter.getReportRule(request), timer);
+        if (log.isLoggable(Level.FINEST)) {
+          log.log(Level.FINEST, String.format("sending an error report request %s", reportRequest));
+        }
+        client.report(reportRequest);
+        statistics.totalFiltered.incrementAndGet();
+        statistics.totalFilteredTime.addAndGet(overallTimer.elapsed(TimeUnit.MILLISECONDS));
+        logStatistics();
+        return;
+      }
+    }
+
+    QuotaRequestInfo quotaInfo = createQuotaInfo(httpRequest, info);
+    if (quotaInfo.getMetricCosts().isEmpty()) {
+      if (log.isLoggable(Level.FINE)) {
```

Comment:
we log at level FINE many other places in this PR without this check; why do we need it here?